supreme_court_predictions/processing/token_aggregation.py

```diff
@@ -0,0 +1,237 @@
+
+        # Merging utterances with voter (judge) and advocate sides
+        ut_sides = pd.merge(
+            ut,
+            pd.concat([self.voteside, self.advocateside]),
+            how="left",
+            left_on=["case_id", "advocate"],
+            right_on=["case_id", "advocate"],
+        )
+
+        # Remove NA values
+        ut_sides = ut_sides.loc[~ut_sides.loc[:, "side"].isna(), :]
+        ut_sides = ut_sides.astype({"side": "int32"})
+
+        return ut_sides
+
+    def get_advocate_case_tokens(self, advocate=True):
+        """
+        Get all of the tokens for individuals either in favor of the petitioner
+        or opposed to the petitioner.
+
+        :param advocate: Whether to find the tokens for those in favor
+                         (advocate=True) or opposed (advocate=False)
+        :return A dataframe of tokens per case for petitioner advocates.
+        """
+
+        if advocate:
+            ut = self.utterance_sides.loc[
+                self.utterance_sides.loc[:, "side"] == 1, :
+            ]
+        else:
+            ut = self.utterance_sides.loc[
+                self.utterance_sides.loc[:, "side"] == 0, :
+            ]
+        return self.get_case_tokens(ut.loc[:, ["case_id", "tokens"]])
+
+    def get_judge_case_tokens(self):
+        """
+        Get all of the tokens for only judges.
+
+        :return A dataframe of tokens per judge per case.
+        """
+
+        ut = self.utterance_sides.loc[
+            self.utterance_sides.loc[:, "speaker_type"] == "J", :
+        ]
+        return self.get_case_tokens(ut.loc[:, ["case_id", "tokens"]])
+
+    def parse_all_data(self):
+        """
+        Generates token aggregations for 1) all speakers, 2) only advocates, 3)
+        only adversaries, 4) only judges, and appends the winning side of the
+        case. DataFrames and exports them to csv/excel (if applicable).
+        """
+
+        print("Grabbing token aggregation for all cases...")
+        self.all_tokens = self.get_all_case_tokens()
+
+        print("Grabbing token aggregation for advocates...")
+        self.advocate_tokens = self.get_advocate_case_tokens(True)
+
+        print("Grabbing token aggregation for adversaries...")
+        self.adversary_tokens = self.get_advocate_case_tokens(False)
+
+        print("Grabbing token aggregation for judges...")
+        self.judge_tokens = self.get_judge_case_tokens()
+
+        if self.save_data:
+            print("Exporting files...")
+            # Outputting to CSVs
+            aggregations = [
+                self.all_tokens,
+                self.advocate_tokens,
+                self.adversary_tokens,
+                self.judge_tokens,
+            ]
+            outpaths = [
+                self.output_path + "/case_aggregations.p",
+                self.output_path + "/advocate_aggregations.p",
+                self.output_path + "/adversary_aggregations.p",
+                self.output_path + "/judge_aggregations.p",
```

Comment:
What is the `.p` extension?

Reply:
It's the extension for pickle files, easier to read and write out to than constantly using CSVs